src/webgpu/api/operation/resource_init/check_texture/by_ds_test.ts

```diff
@@ -0,0 +1,197 @@
+import { assert } from '../../../../../common/framework/util/util.js';
+import { GPUTest } from '../../../../gpu_test.js';
+import { CheckContents } from '../texture_zero_init.spec.js';
+
+function fullscreenVertexModule(device: GPUDevice) {
+  return device.createShaderModule({
+    code: `
+    [[builtin(position)]] var<out> Position : vec4<f32>;
+    [[builtin(vertex_idx)]] var<in> VertexIndex : i32;
+
+    [[stage(vertex)]]
+    fn main() -> void {
+      const pos : array<vec2<f32>, 3> = array<vec2<f32>, 3>(
+        vec2<f32>(-1.0, -3.0),
+        vec2<f32>( 3.0,  1.0),
+        vec2<f32>(-1.0,  1.0));
+      Position = vec4<f32>(pos[VertexIndex], 0.0, 1.0);
+      return;
+    }
+    `,
+  });
+}
+
+function getDepthTestReadbackPipeline(
+  t: GPUTest,
+  format: GPUTextureFormat,
+  sampleCount: number,
+  expected: number
+): GPURenderPipeline {
+  return t.device.createRenderPipeline({
+    vertexStage: {
+      entryPoint: 'main',
+      module: fullscreenVertexModule(t.device),
+    },
+    fragmentStage: {
+      entryPoint: 'main',
+      module: t.device.createShaderModule({
+        code: `
+        [[builtin(frag_depth)]] var<out> FragDepth : f32;
+        [[location(0)]] var<out> outSuccess : f32;
+
+        [[stage(fragment)]]
+        fn main() -> void {
+          FragDepth = f32(${expected});
+          outSuccess = 1.0;
+          return;
+        }
+        `,
+      }),
+    },
+    colorStates: [
+      {
+        format: 'r8unorm',
+      },
+    ],
+    depthStencilState: {
+      format,
+      depthCompare: 'equal',
+    },
+    primitiveTopology: 'triangle-list',
+    sampleCount,
+  });
+}
+
+function getStencilTestReadbackPipeline(
+  t: GPUTest,
+  format: GPUTextureFormat,
+  sampleCount: number
+): GPURenderPipeline {
+  return t.device.createRenderPipeline({
+    vertexStage: {
+      entryPoint: 'main',
+      module: fullscreenVertexModule(t.device),
+    },
+    fragmentStage: {
+      entryPoint: 'main',
+      module: t.device.createShaderModule({
+        code: `
+        [[location(0)]] var<out> outSuccess : f32;
+
+        [[stage(fragment)]]
+        fn main() -> void {
+          outSuccess = 1.0;
+          return;
+        }
+        `,
+      }),
+    },
+    colorStates: [
+      {
+        format: 'r8unorm',
+      },
+    ],
+    depthStencilState: {
+      format,
+      stencilFront: {
+        compare: 'equal',
+      },
+      stencilBack: {
+        compare: 'equal',
+      },
+    },
+    primitiveTopology: 'triangle-list',
+    sampleCount,
+  });
+}
+
+const checkContents: (type: 'depth' | 'stencil', ...args: Parameters<CheckContents>) => void = (
+  type,
+  t,
+  params,
+  texture,
+  state,
+  subresourceRange
+) => {
+  for (const viewDescriptor of t.generateTextureViewDescriptorsForRendering(
+    params.aspect,
+    subresourceRange
+  )) {
+    const width = t.textureWidth >> viewDescriptor.baseMipLevel!;
```

Comment:
Shouldn't this be using a helper function to do the max(1, ...) ? ditto in other files.